CIP-XXXX/CIP-XXXX.md

```diff
@@ -0,0 +1,558 @@
+bitwiseShift [] (-3) => []
+-- Positive shifts move bits to higher indexes, cutting off high indexes and
+-- filling low ones with zeroes
+bitwiseShift [0xEB, 0xFC] 5 => [0x7F, 0x80]
+-- Negative shifts move bits to lower indexes, cutting off low indexes and
+-- filling high ones with zeroes
+bitwiseShift [0xEB, 0xFC] (-5) => [0x07, 0x5F]
+-- Shifting by the total number of bits or more clears all bytes
+bitwiseShift [0xEB, 0xFC] 16 => [0x00, 0x00]
+-- Regardless of direction
+bitwiseShift [0xEB, 0xFC] (-16) => [0x00, 0x00]
+```
+
+#### `bitwiseRotate`
+
+`bitwiseRotate` takes two arguments; we name and describe them below.
+
+1. The `BuiltinByteString` to be rotated. This is the _data argument_.
+2. The rotation, whose sign indicates direction and whose magnitude indicates 
+   the size of the rotation. This is the _rotation argument_, and has type
+   `BuiltinInteger`.
+
+Let $b$ refer to the data argument, whose length in bytes is $n$, and let $i$
+refer to the rotation argument. Let the result of `bitwiseRotate` called with $b$
+and $i$ be $b_r$, also of length $n$. 
+
+For all $j \in 0, 1, \ldots 8 \cdot n - 1$, we have $b_r = b[j - i \mod 8 \cdot n]$.
+
+Some examples of the intended behaviour of `bitwiseRotate` follow. For
+brevity, we write `BuiltinByteString` literals as lists of hexadecimal values.
+
+```
+-- Rotating the empty bytestring does nothing
+bitwiseRotate [] 3 => []
+-- Regardless of direction
+bitwiseRotate [] (-1) => []
+-- Positive rotations move bits to higher indexes, 'wrapping around' for high
+-- indexes into low indexes
+bitwiseRotate [0xEB, 0xFC] 5 => [0x7F, 0x9D]
+-- Negative rotations move bits to lower indexes, 'wrapping around' for low
+-- indexes into high indexes
+bitwiseRotate [0xEB, 0xFC] (-5) => [0xE7, 0x5F]
+-- Rotation by the total number of bits does nothing
+bitwiseRotate [0xEB, 0xFC] 16 => [0xEB, 0xFC]
+-- Regardless of direction
+bitwiseRotate [0xEB, 0xFC] (-16) => [0xEB, 0xFC]
+-- Rotation by more than the total number of bits is the same as the remainder
+-- after division by number of bits
+bitwiseRotate [0xEB, 0xFC] 21 =>[0x7F, 0x9D]
+-- Regardless of direction, preserving sign
+bitwiseRotate [0xEB, 0xFC] (-21) => [0xE7, 0x5F]
+```
+
+#### `countSetBits`
+
+Let $b$ refer to `countSetBits`' only argument, whose length in bytes is $n$,
+and let $r$ be the result of calling `countSetBits` on $b$. Then we have
+
+$$
+r = \sum_{i=0}^{8 \cdot n - 1} b[i]
+$$
+
+Some examples of the intended behaviour of `countSetBits` follow. For
+brevity, we write `BuiltinByteString` literals as lists of hexadecimal values.
+
+```
+-- The empty bytestring has no set bits
+countSetBits [] => 0
+-- Bytestrings with only zero bytes have no set bits
+countSetBits [0x00, 0x00] => 0
+-- Set bits are counted regardless of where they are
+countSetBits [0x01, 0x00] => 1
+countSetBits [0x00, 0x01] => 1
+```
+
+#### `findFirstSetBit`
+
+Let $b$ refer to `findFirstSetBit`'s only argument, whose length in bytes is $n$,
```

Comment:
I suppose there's a slight danger of confusion here because we index bits in bytestrings from the right (which I agree is a good choice) but if you write down the bytestring in some form such as `[0x1a, 0x35]` or `0x1a35` or `(con bytestring #1a35)` this function will actually find what appears to be the _last_ bit set.  I think that's an  unavoidable consequence of the way we write things down though (and the cultural fact that we read from left to right), and I can't think of a more sensible name.  I guess we should just make it clear how it works in the documentation.

Reply:
Yeah, 'first' here means 'based on our indexing scheme'. This whole left-versus-right situation is quite unfortunate, but as CIP-122 explains, unavoidable.